File: crates/issundb-py/examples/quickstart.py
import os
import shutil
import json
from issundb import IssunDB

def main():
    db_path = "./issundb-py-quickstart-data"

    # Clean up from previous run if any
    if os.path.exists(db_path):
        shutil.rmtree(db_path)

    print("IssunDB Python Quickstart Example")
    print("================================")

    # 1. Open the database
    db = IssunDB(db_path)
    print(f"Opened database at: {db_path}")

    # 2. Add some nodes
    alice_props = json.dumps({"name": "Alice", "age": 30})
    alice_id = db.add_node("Person", alice_props)
    print(f"Created node Alice with ID: {alice_id}")

    bob_props = json.dumps({"name": "Bob", "age": 25})
    bob_id = db.add_node("Person", bob_props)
    print(f"Created node Bob with ID: {bob_id}")

    # 3. Connect them with an edge
    edge_props = json.dumps({"since": 2021})
    edge_id = db.add_edge(alice_id, bob_id, "KNOWS", edge_props)
    print(f"Created KNOWS edge with ID: {edge_id}")

    # 4. Query using Cypher
    cypher_query = "MATCH (a:Person)-[r:KNOWS]->(b:Person) RETURN a.name, b.name, r.since"
    print(f"\nExecuting query: {cypher_query}")

    result_str = db.query(cypher_query)
    result = json.loads(result_str)

    print("\nQuery results:")
    print("Columns:", result["columns"])
    for record in result["records"]:
        values = record["values"]
        print(f"  - {values[0]} knows {values[1]} since {values[2]}")

    # 5. Explain query execution plan
    print("\nQuery Explanation Plan:")
    plan = db.explain(cypher_query)
    print(plan)

    # Clean up database files
    if os.path.exists(db_path):
        shutil.rmtree(db_path)
    print("Database cleaned up successfully.")

if __name__ == "__main__":
    main()

File: crates/issundb-examples/quickstart.rs
use issundb::{Graph, GraphQueryExt};
use std::path::Path;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Clean up any existing data directory from a previous run to start from a clean slate
    let db_path = Path::new("./issundb-data-quickstart");
    if db_path.exists() {
        let _ = std::fs::remove_dir_all(db_path);
    }

    // Open a graph database with a 1 GB memory map size limit
    let graph = Graph::open(db_path, 1)?;

    // Add two nodes with properties
    let alice_props = serde_json::json!({ "name": "Alice", "age": 30 });
    let alice_id = graph.add_node("Person", &alice_props)?;

    // Add another node
    let bob_props = serde_json::json!({ "name": "Bob", "age": 28 });
    let bob_id = graph.add_node("Person", &bob_props)?;

    // Create an edge between Alice and Bob with a property
    let edge_props = serde_json::json!({ "since": 2021 });
    graph.add_edge(alice_id, bob_id, "KNOWS", &edge_props)?;

    // Optional: rebuild CSR snapshot manually after bulk writes
    graph.rebuild_csr()?;

    // Execute a Cypher query and print the results
    let result =
        graph.query("MATCH (a:Person)-[r:KNOWS]->(b:Person) RETURN a.name, b.name, r.since")?;

    for record in result.records {
        println!(
            "Match: {} knows {} since {}",
            record.values[0], record.values[1], record.values[2]
        );
    }

    // Remove the database directory now that we're done with the example
    if db_path.exists() {
        let _ = std::fs::remove_dir_all(db_path);
    }

    Ok(())
}
